go/src/koding/kites/kloud/provider/google/stack.go

```diff
@@ -0,0 +1,254 @@
+	_ stack.Stacker  = (*Stack)(nil) // internal API
+)
+
+func newStack(bs *provider.BaseStack) (provider.Stack, error) {
+	return &Stack{BaseStack: bs}, nil
+}
+
+func (s *Stack) VerifyCredential(c *stack.Credential) error {
+	cred := c.Credential.(*Cred)
+
+	if err := cred.Valid(); err != nil {
+		return err
+	}
+
+	computeService, err := cred.ComputeService()
+	if err != nil {
+		return err
+	}
+
+	// Try to get project info. If there is no error, project name and JSON key
+	// are valid.
+	_, err = compute.NewProjectsService(computeService).Get(cred.Project).Do()
+	return err
+}
+
+func (s *Stack) BootstrapTemplates(c *stack.Credential) ([]*stack.Template, error) {
+	cfg := &bootstrapConfig{
+		Username:  s.Req.Username,
+		PublicKey: s.Keys.PublicKey,
+	}
+
+	t, err := newBootstrapTemplate(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	cred := c.Credential.(*Cred)
+	t.Key = cred.Project + "-" + s.BootstrapArg().GroupName + "-" + c.Identifier
+	return []*stack.Template{t}, nil
+}
+
+func (s *Stack) ApplyTemplate(c *stack.Credential) (*stack.Template, error) {
+	t := s.Builder.Template
+
+	var resource struct {
+		GCInstance map[string]map[string]interface{} `hcl:"google_compute_instance"`
+	}
+
+	if err := t.DecodeResource(&resource); err != nil {
+		return nil, err
+	}
+
+	if len(resource.GCInstance) == 0 {
+		return nil, fmt.Errorf("there are no Google compute instances defined")
+	}
+
+	for resourceName, instance := range resource.GCInstance {
+		// Set default machine type if user didn't define it herself.
+		if mt, ok := instance["machine_type"]; !ok {
+			instance["machine_type"] = defaultMachineType
+		} else {
+			if mtstr, ok := mt.(string); ok && mtstr == "" {
+				instance["machine_type"] = defaultMachineType
+			}
+		}
+
+		// Set default zone if user didn't define it herself.
+		const lookupZone = "${lookup(var.zones, var.google_region)}"
+		if z, ok := instance["zone"]; !ok {
+			instance["zone"] = lookupZone
+		} else {
+			if zstr, ok := z.(string); ok && zstr == "" {
+				instance["zone"] = lookupZone
+			}
+		}
+
+		// Set default image for disk if user didn't define it herself.
+		if _, ok := instance["disk"]; !ok {
+			instance["disk"] = struct {
+				Image string `json:"image" bson:"image" hcl:"image"`
+			}{
+				Image: defualtMachineImage,
+			}
+		}
+
+		// Set default network interface if user didn't define it herself.
+		if _, ok := instance["network_interface"]; !ok {
+			instance["network_interface"] = struct {
+				Network      string   `json:"network" bson:"network" hcl:"network"`
+				AccessConfig struct{} `json:"access_config" bson:"access_config" hcl:"access_config"`
+			}{
+				Network:      "default",
+				AccessConfig: struct{}{},
+			}
+		}
+
+		// means there will be several instances, we need to create a userdata
+		// with count interpolation, because each machine must have an unique
+		// kite id.
+		count := 1
+		if n, ok := instance["count"].(int); ok && n > 1 {
```

Comment:
google provider does not support count attribute [0] like AWS one does, so there's no need to generate multiple keys for a single compute_instance resource (the ApplyTemplate will look more like vagrant than aws kloud provider) 

[0] https://www.terraform.io/docs/providers/google/r/compute_instance.html